# Retail Electric Service Territories
Explore the shape and content of the [service territories map from HIFLD](https://hifld-geoplatform.opendata.arcgis.com/datasets/electric-retail-service-territories-2/explore?location=43.479127%2C-92.210913%2C3.79)

In [1]:
import geopandas as gpd
import pandas as pd
from pathlib import Path

/app/.local/lib/python3.9/site-packages/geopandas/_compat.py:111: UserWarning: The Shapely GEOS version (3.10.3-CAPI-1.16.1) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.
  warnings.warn(


In [2]:
shapefile = Path('/app/data/data_cache/Electric_Retail_Service_Territories.zip')
assert shapefile.exists()

In [3]:
# Read in data
territories = gpd.read_file(shapefile)

In [4]:
territories.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 2949 entries, 0 to 2948
Data columns (total 37 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   OBJECTID    2949 non-null   int64   
 1   ID          2949 non-null   object  
 2   NAME        2949 non-null   object  
 3   ADDRESS     2949 non-null   object  
 4   CITY        2949 non-null   object  
 5   STATE       2949 non-null   object  
 6   ZIP         2949 non-null   object  
 7   TELEPHONE   2949 non-null   object  
 8   TYPE        2949 non-null   object  
 9   COUNTRY     2949 non-null   object  
 10  NAICS_CODE  2949 non-null   object  
 11  NAICS_DESC  2949 non-null   object  
 12  SOURCE      2949 non-null   object  
 13  SOURCEDATE  2949 non-null   object  
 14  VAL_METHOD  2949 non-null   object  
 15  VAL_DATE    2949 non-null   object  
 16  WEBSITE     2949 non-null   object  
 17  REGULATED   2949 non-null   object  
 18  CNTRL_AREA  2949 non-null   object  
 19

In [5]:
pd.options.display.max_columns = 50

In [6]:
territories.head()

,OBJECTID,ID,NAME,ADDRESS,CITY,STATE,ZIP,TELEPHONE,TYPE,COUNTRY,NAICS_CODE,NAICS_DESC,SOURCE,SOURCEDATE,VAL_METHOD,VAL_DATE,WEBSITE,REGULATED,CNTRL_AREA,PLAN_AREA,HOLDING_CO,SUMMR_PEAK,WINTR_PEAK,SUMMER_CAP,WINTER_CAP,NET_GEN,PURCHASED,NET_EX,RETAIL_MWH,WSALE_MWH,TOTAL_MWH,TRANS_MWH,CUSTOMERS,YEAR,Shape__Are,Shape__Len,geometry
0,1,1000,CITY OF AUGUSTA - (AR),NOT AVAILABLE,NOT AVAILABLE,AR,NOT AVAILABLE,NOT AVAILABLE,MUNICIPAL,USA,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...","EIA 861, TIGER/Line Shapefiles - U.S. Census",2021-10-06,OTHER,2018-10-12,NOT AVAILABLE,NOT AVAILABLE,MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...,NOT AVAILABLE,CITY OF AUGUSTA - (AR),-999999.0,-999999.0,-999999.0,-999999.0,12823,-999999,-999999,-999999,-999999,-999999,-999999,916,2020,0.152271,2.625553,"POLYGON ((-91.34713 35.43899, -91.34711 35.439..."
1,2,10000,EVERGY METRO,1200 MAIN STREET,KANSAS CITY,KS,64105,NOT AVAILABLE,INVESTOR OWNED,USA,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...",https://www.kcpl.com/-/media/indexedmedia/abou...,2021-10-06,OTHER,2021-04-30,https://www.evergy.com/,NOT AVAILABLE,SOUTHWEST POWER POOL,NOT AVAILABLE,KANSAS CITY POWER & LIGHT CO,3200.0,2484.0,4877.8,-999999.0,14911882,5624474,0,14223892,5957441,20536356,0,562180,2020,0.525845,10.131485,"MULTIPOLYGON (((-94.66607 38.27027, -94.66941 ..."
2,3,10005,"EVERGY KANSAS SOUTH, INC",818 KANSAS AVENUE,TOPEKA,KS,66612,NOT AVAILABLE,INVESTOR OWNED,USA,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...","EIA 860, EIA 861, TIGER/Line Shapefiles - U.S....",2021-10-06,OTHER,2021-04-30,https://www.evergy.com/,REGULATED,SOUTHWEST POWER POOL,NOT AVAILABLE,WESTAR ENERGY,2135.0,1421.0,295.1,295.1,9324681,1566609,0,9181326,1628176,10891290,0,334500,2020,3.937805,15.439905,"POLYGON ((-95.50819 38.42458, -95.50818 38.415..."
3,4,10009,KARNES ELECTRIC COOP INC,1007 N. HWY. 123,KARNES CITY,TX,78118,(830) 780-3952,COOPERATIVE,USA,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...","EIA 861, TIGER/Line Shapefiles - U.S. Census",2021-10-06,OTHER,2018-10-12,http://www.karnesec.org/,NOT AVAILABLE,"ELECTRIC RELIABILITY COUNCIL OF TEXAS, INC.",ERCOT,KARNES ELECTRIC COOP INC,207.0,208.0,-999999.0,-999999.0,-999999,1170464,0,1129899,-999999,1170464,0,21327,2020,3.085930,8.976781,"POLYGON ((-98.54256 29.75188, -98.54102 29.751..."
4,5,10012,KAY ELECTRIC COOP,300 W. DOOLIN,BLACKWELL,OK,74631,(580) 363-1260,COOPERATIVE,USA,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...","EIA 861, TIGER/Line Shapefiles - U.S. Census, ...",2021-10-06,OTHER,2019-05-01,www.kayelectric.coop,NOT AVAILABLE,SOUTHWEST POWER POOL,WESTERN FARMERS ELECTRIC COOPERATIVE,KAY ELECTRIC COOP,71.0,71.0,-999999.0,-999999.0,0,378230,0,371371,0,378230,0,5785,2020,0.599957,3.978050,"POLYGON ((-97.74419 36.93999, -97.74638 36.977..."


Confirmed that the `ID` field is the EIA utility ID, at least for US utilities. Querying PUDL shows:
```-- SQLite
SELECT
    utility_id_eia,
    utility_name_eia
FROM utilities_entity_eia
WHERE utility_name_eia LIKE 'evergy%'
```

|utility_id_eia|utility_name_eia|
|---|---|
|10000|Evergy Metro|
|10005|Evergy Kansas South, Inc|
|56211|Evergy Missouri West|
|64428|Evergy, Inc.|


In [16]:
territories['ID'].is_unique

True

In [17]:
territories.query("COUNTRY != 'USA'").head()

,OBJECTID,ID,NAME,ADDRESS,CITY,STATE,ZIP,TELEPHONE,TYPE,COUNTRY,NAICS_CODE,NAICS_DESC,SOURCE,SOURCEDATE,VAL_METHOD,VAL_DATE,WEBSITE,REGULATED,CNTRL_AREA,PLAN_AREA,HOLDING_CO,SUMMR_PEAK,WINTR_PEAK,SUMMER_CAP,WINTER_CAP,NET_GEN,PURCHASED,NET_EX,RETAIL_MWH,WSALE_MWH,TOTAL_MWH,TRANS_MWH,CUSTOMERS,YEAR,Shape__Are,Shape__Len,geometry
2793,2794,NA1,FORTISBC,"SUITE 100, 1975 SPRINGFIELD ROAD",KELOWNA,BC,V1Y 7V7,(866) 436-7847,INVESTOR OWNED,CAN,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...",http://www.atcoelectric.com/About-Us/Our-Servi...,2015-09-30,OTHER,2020-05-18,https://outages.fortisbc.com/Outages/Home/Outa...,NOT AVAILABLE,NOT AVAILABLE,NOT AVAILABLE,FORTISBC,0.0,0.0,-999999.0,-999999.0,-999999,-999999,-999999,-999999,-999999,-999999,-999999,-999999,NOT AVAILABLE,2.804753,10.264503,"MULTIPOLYGON (((-117.22049 49.97579, -117.2272..."
2794,2795,NA10,MEDICINE HAT (CITY OF),2172 BRIER PARK PLACE NW,MEDICINE HAT,AB,T1C 1S6,(403) 529-8262,MUNICIPAL,CAN,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...",http://www.atcoelectric.com/About-Us/Our-Servi...,2015-09-30,OTHER,2015-09-30,http://www.city-medicine-hat-ab-ca/,NOT AVAILABLE,NOT AVAILABLE,NOT AVAILABLE,MEDICINE HAT (CITY OF),0.0,0.0,-999999.0,-999999.0,-999999,-999999,-999999,-999999,-999999,-999999,-999999,-999999,NOT AVAILABLE,0.026733,0.739826,"POLYGON ((-110.61980 50.05340, -110.61842 49.9..."
2795,2796,NA11,LETHBRIDGE ELECTRIC UTILITY(CITY OF),CITY OF LETHBRIDGE 910 . 4TH AVENUE SOUTH,LETHBRIDGE,AB,T1J 0P6,(403) 320-3940,MUNICIPAL,CAN,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...",http://www.atcoelectric.com/About-Us/Our-Servi...,2015-09-30,OTHER,2015-09-30,http://www.lethbridge-ca/,NOT AVAILABLE,NOT AVAILABLE,NOT AVAILABLE,LETHBRIDGE ELECTRIC UTILITY(CITY OF),0.0,0.0,-999999.0,-999999.0,-999999,-999999,-999999,-999999,-999999,-999999,-999999,-999999,NOT AVAILABLE,0.016966,0.670017,"POLYGON ((-112.81188 49.75366, -112.81732 49.6..."
2796,2797,NA12,SILVERSMITH POWER & LIGHT CORP,SILVERSMITH POWER & LIGHT CORPORATION BOX 369,NEW DENVER,BC,V0G 1S0,(250) 358-2247,NOT AVAILABLE,CAN,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...",http://canadiangis.com/data.php,2015-09-30,OTHER,2015-09-30,http://www.sandonbc-ca/,NOT AVAILABLE,NOT AVAILABLE,NOT AVAILABLE,SILVERSMITH POWER & LIGHT CORP,0.0,0.0,-999999.0,-999999.0,-999999,-999999,-999999,-999999,-999999,-999999,-999999,-999999,NOT AVAILABLE,0.000052,0.060308,"POLYGON ((-117.22156 49.97748, -117.22049 49.9..."
2799,2800,NA2,BC HYDRO,"1155 MCGILL RD,",KAMLOOPS,BC,V2C,(888) 769-3766,FEDERAL,CAN,2211,"ELECTRIC POWER GENERATION, TRANSMISSION AND DI...",http://www.arcgis.com/home/item.html?id=dcbcdf...,2015-09-30,OTHER,2015-09-30,https://www.bchydro.com/contact.html,NOT AVAILABLE,NOT AVAILABLE,NOT AVAILABLE,BC HYDRO,0.0,0.0,-999999.0,-999999.0,-999999,-999999,-999999,-999999,-999999,-999999,-999999,-999999,NOT AVAILABLE,128.531030,197.229611,"MULTIPOLYGON (((-125.00080 60.00387, -123.8169..."


In [7]:
import numpy as np

In [8]:
territories.replace([-999999.0, ''], np.nan).info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 2949 entries, 0 to 2948
Data columns (total 37 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   OBJECTID    2949 non-null   float64 
 1   ID          2949 non-null   object  
 2   NAME        2949 non-null   object  
 3   ADDRESS     2949 non-null   object  
 4   CITY        2949 non-null   object  
 5   STATE       2949 non-null   object  
 6   ZIP         2949 non-null   object  
 7   TELEPHONE   2949 non-null   object  
 8   TYPE        2949 non-null   object  
 9   COUNTRY     2949 non-null   object  
 10  NAICS_CODE  2949 non-null   object  
 11  NAICS_DESC  2949 non-null   object  
 12  SOURCE      2949 non-null   object  
 13  SOURCEDATE  2949 non-null   object  
 14  VAL_METHOD  2949 non-null   object  
 15  VAL_DATE    2949 non-null   object  
 16  WEBSITE     2949 non-null   object  
 17  REGULATED   2949 non-null   object  
 18  CNTRL_AREA  2949 non-null   object  
 19

In [11]:
territories['STATE'].nunique()

58

In [13]:
territories['STATE'].value_counts(dropna=False, sort=False).sort_index()

AB      7
AK     65
AL     59
AR     34
AS      1
AZ     38
BC      5
CA     54
CO     54
CT     11
DC      1
DE     11
FL     53
GA     94
GU      1
HI      4
IA    175
ID     26
IL     69
IN    114
KS    149
KY     54
LA     37
MA     44
MD      8
ME      9
MI     55
MN    170
MO    128
MP      1
MS     50
MT     25
NC    101
ND     31
NE    153
NH     10
NJ     13
NM     24
NV     13
NY     62
OH    125
OK     91
OR     36
PA     58
PR      1
RI      3
SC     44
SD     62
TN     82
TX    140
UT     43
VA     29
VI      1
VT     17
WA     59
WI    117
WV      4
WY     24
Name: STATE, dtype: int64